### DISCUSSION TITLE
zipEqual, zipLongest

### DISCUSSION THREAD

**Discussion by cjquines** [remeda/remeda#955] [Suggest a Function]
Two related but separate proposals:

- `zipEqual`, which would be the equivalent of [`zip(..., strict=True)` from Python](https://peps.python.org/pep-0618/) or [`itertools::zip_eq` from Rust](https://docs.rs/itertools/latest/itertools/fn.zip_eq.html).
- `zipLongest`, which would be the equivalent of [`itertools.zip_longest` from Python](https://docs.python.org/3.11/library/itertools.html#itertools.zip_longest) and the approximate equivalent of [`transpose` from Ramda](https://ramdajs.com/docs/#transpose).
    - Options include adding a fill element like `itertools.zip_longest`, or omitting like `transpose`.

**Comment by eranhirsch** [comment]
I'm not sure these add enough, the user would have to read several function's documentation just to make sure they are using the correct version. I think that our current zip implements the most expected version of the function as is common in other libraries, and if users need their own variant they can always write their own utility.

I do think that we should add an `invariant` function that can throw on a provided predicate similar to an inline invariant. Having that would allow users to implement zipEqual via 2 lines of a pipe:
```ts
const x = pipe(
  ...
  invariant(isLength(other.length)),
  zip(other),
  ...
);
```

**Comment by eranhirsch** [comment]
@TkDodo, wdyt?

**Comment by cjquines** [comment]
`invariant` would be interesting, especially if you could use it with type guards:

```ts
const data: number | string = "a";
const x = R.pipe(
  data,
  R.invariant(R.isNumber),
  R.add(2), // ok
);
```

why not the name `assert`?

**Comment by eranhirsch** [comment]
I think I tried using assert in the past and encountered some problem, but I don't remember the specifics and agree it could be called `assert` if it works.

**Comment by Powell-v2** [comment]
hi @eranhirsch, I'd like to lend a hand with this one - have you already decided on the desired implementation strategy?

**Comment by eranhirsch** [comment]
I don't believe this request is general enough to grant adding to the library. Others are asking if we can add a zip function that would take an arbitrary number of arrays instead of just 2. Together with that request, the number of zip functions we will end up supporting would be too big because of limitations to how Remeda can support optional params and kwargs-style params, which it simply can't because of our runtime currying; it requires every variant to be a top-level util.

These utils could be implemented on top of Remeda with a short helper function of 5-10 lines of code.

I'll leave the issue open so others can comment, but for now, I don't support adding these.

**Comment by Powell-v2** [comment]
Thanks for the update! Fair enough. Which areas of the library could use some help from your point of view? Happy to begin contributing to the project : )

**Comment by eranhirsch** [comment]
There are a few functions that need more test coverage. Tests are always a good addition because they allow us to ensure we continue delivering a high-quality library while we change, modernize, and fix stuff. This is especially useful for us going into V2 as we are going to change, modernize, and fix quite a few things 😉

**Comment by Powell-v2** [comment]
Of course, testing is foundational : ) Could you please give me some pointers?

**Comment by eranhirsch** [comment]
Any test file that is short, lacks interesting cases, or has an issue in the testing code itself. Just notice that we are deprecating some functions, so avoid those test files (see the @deprecated tag in the function impl).
